chore(blog): add API Drift in Event-Driven Architectures and AsyncAPI Studio posts (#5405)
asyncapi/asyncapi.github.io: markdown/blog/api-drift-in-event-driven-architectures-has-two-faces.md

The broker and schema infrastructure become part of the behavioural model of the system itself.

This changes what API Drift means in Event-Driven Architectures.

API Drift in Event-Driven Architectures is a problem with two faces.

---

## Implementation Drift

This is the drift most teams already know:
- generated code diverges
- implementations evolve separately
- consumers rely on undocumented behaviour
- contracts and services fall out of sync

API-First development practices solve this increasingly well.

Using AsyncAPI as the canonical representation, we can:
- generate code
- validate schemas
- continuously rebuild implementations from the contract

This is API-First Dev. The contract becomes the source of truth for implementation.

We can reduce implementation drift by generating read-only code on every build.

If generated code is not committed to source control, and implementations are continuously rebuilt
from the canonical API contract, specification and implementation stay synchronized by construction.

> ---

Reviewer: ```suggestion
```

## Platform Drift

In Event-Driven Architectures, infrastructure itself also drifts.

Over time:
- topic configuration diverges
- retention policies evolve independently
- ACLs drift
- infrastructure and services make different assumptions

Eventually:
- the contract says one thing
- the platform runs another
- applications assume something else

This is platform drift.

And unlike implementation drift, most API-First approaches still ignore it.

> But if the broker shapes system behavior, then infrastructure itself must become contract-driven.

Reviewer: ```suggestion
But if the broker shapes system behaviour, then infrastructure itself must become contract-driven.
```

This is where API-First Ops begins.

Using AsyncAPI as the operational source of truth, we can derive:
- Kafka topics
- Schema Registry subjects
- ACLs
- Terraform infrastructure
- governance validation

directly from the API contract itself.

One contract. Two faces of API Drift solved together.

> ---

Reviewer: ```suggestion
```

## Tooling

These are not new ideas.

The implementation side has been running in production Java and Spring shops since 2020. It started

> as a closed source internal tool inside a large organization and became open source in 2023. The

Reviewer: ```suggestion
as a closed-source internal tool inside a large organisation, and became open source in 2023. The
```

> AsyncAPI Generator for Spring Boot and Spring Cloud Stream and Spring Kafka is the direct evolution of that work. It

Reviewer: ```suggestion
AsyncAPI Generator for Spring Boot, Spring Cloud Stream, and Spring Kafka is the direct evolution of that work. It
```

---
title: API Drift in Event-Driven Architectures has Two Faces
date: 2026-05-15T09:00:00+02:00
type: Engineering
tags:
  - API-First
  - API-First Ops
  - AsyncAPI
  - Event-Driven Architecture
  - Kafka
  - Terraform
  - Spring Boot
  - Platform Engineering
  - API Governance
  - API Drift
  - ZenWave SDK
cover: /img/posts/zenwave/api-drift-two-faces-banner.png
authors:
  - name: Ivan Garcia Sainz-Aja
    photo: /img/avatars/ivangsa.webp
    link: https://ivangsa.com/about
    byline: Software Engineer. Speaker. Building ZenWave 360º
excerpt: 'In Event-Driven Architectures, API Drift happens in two places: implementation and platform. AsyncAPI and API-First Ops can keep both continuously aligned from a single contract.'
---

> # API Drift in Event-Driven Architectures has Two Faces

Reviewer: ```suggestion

```

Reviewer: Let're remove the heading since its already in the title

> In REST architectures, gateways usually mediate communication. They handle routing, authentication, observability, and rate limiting, but they rarely shape the behavior of the system itself.

Reviewer: ```suggestion
In REST architectures, gateways usually mediate communication. They handle routing, authentication, observability, and rate limiting, but they rarely shape the system's behaviour.
```

Event-Driven Architectures are different.

> In EDA, the broker shapes behavioral semantics. Kafka topics define scalability, ordering guarantees, replay capabilities, and retention behavior. Schema Registries define schema evolution, compatibility rules, and message contracts.

Reviewer: ```suggestion
In EDA, the broker shapes behavioural semantics. Kafka topics define scalability, ordering guarantees, replay capabilities, and retention behaviour. Schema Registries define schema evolution, compatibility rules, and message contracts.
```

> The broker and schema infrastructure become part of the behavioral model of the system itself.

Reviewer: ```suggestion
The broker and schema infrastructure become part of the behavioural model of the system itself.
```

This changes what API Drift means in Event-Driven Architectures.

API Drift in Event-Driven Architectures is a problem with two faces.

> ---

Reviewer: ```suggestion
```

## Implementation Drift

This is the drift most teams already know:
- generated code diverges
- implementations evolve separately

> - consumers rely on undocumented behavior

Reviewer: ```suggestion
- consumers rely on undocumented behaviour
```

- contracts and services fall out of sync

API-First development practices solve this increasingly well.

Using AsyncAPI as the canonical representation, we can:
- generate code
- validate schemas
- continuously rebuild implementations from the contract

This is API-First Dev. The contract becomes the source of truth for implementation.

We can reduce implementation drift by generating read-only code on every build.

> If generated code is not committed to source control, and implementations are continuously rebuilt

Reviewer: ```suggestion
If the generated code is not committed to source control, and implementations are continuously rebuilt
```
